hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/file/impl/LocalFileSourceFactory.java

```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jet.pipeline.file.impl;
+
+import com.fasterxml.jackson.databind.ObjectReader;
+import com.fasterxml.jackson.dataformat.csv.CsvMapper;
+import com.fasterxml.jackson.dataformat.csv.CsvSchema;
+import com.hazelcast.function.FunctionEx;
+import com.hazelcast.jet.datamodel.Tuple2;
+import com.hazelcast.jet.impl.util.IOUtil;
+import com.hazelcast.jet.json.JsonUtil;
+import com.hazelcast.jet.pipeline.BatchSource;
+import com.hazelcast.jet.pipeline.Sources;
+import com.hazelcast.jet.pipeline.file.CsvFileFormat;
+import com.hazelcast.jet.pipeline.file.FileFormat;
+import com.hazelcast.jet.pipeline.file.FileSourceBuilder;
+import com.hazelcast.jet.pipeline.file.FileSourceFactory;
+import com.hazelcast.jet.pipeline.file.JsonFileFormat;
+import com.hazelcast.jet.pipeline.file.LinesTextFileFormat;
+import com.hazelcast.jet.pipeline.file.ParquetFileFormat;
+import com.hazelcast.jet.pipeline.file.RawBytesFileFormat;
+import com.hazelcast.jet.pipeline.file.TextFileFormat;
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+import java.io.BufferedReader;
+import java.io.FileInputStream;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.nio.charset.Charset;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.ServiceLoader;
+import java.util.Spliterator;
+import java.util.Spliterators;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+import static com.hazelcast.jet.datamodel.Tuple2.tuple2;
+import static com.hazelcast.jet.impl.util.Util.uncheckCall;
+import static com.hazelcast.jet.impl.util.Util.uncheckRun;
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+/**
+ * Implementation of FileSourceFactory for local filesystem
+ *
+ * @param <T> type of the item emitted from the source
+ */
+public class LocalFileSourceFactory<T> implements FileSourceFactory<T> {
+
+    private Map<String, MapFnProvider<? extends FileFormat<?>, ?>> mapFns;
+
+    /**
+     * Default constructor
+     */
+    public LocalFileSourceFactory() {
+        mapFns = new HashMap<>();
+
+        mapFns.put("csv", new CsvMapFnProvider());
```

Comment:
Get the key from the provider? To keep them in sync?

Reply:
Done.